grace-core/src/Grace/Interpret.hs

```diff
@@ -72,31 +69,56 @@ interpretWith
     -> m (Type Location, Value)
 interpretWith bindings maybeAnnotation input = do
-    code <- case input of
-        Path file -> liftIO (Text.IO.readFile file)
-        Code text -> return text
+    let directory = case input of
+            Path file -> FilePath.takeDirectory file
+            Code _ _  -> "."
 
-    let name = case input of
-            Path file -> file
-            Code _    -> "(input)"
+    eitherExpression <- Except.runExceptT (Parser.parseInput input)
 
-    expression <- case Parser.parse name code of
-        Left message -> do
-            Except.throwError (ParseError message)
+    expression <- case eitherExpression of
+        Left e -> throwError (ParseError e)
+        Right expression -> return expression
 
-        Right expression -> do
-            let locate offset = Location{..}
+    interpretExprWith bindings maybeAnnotation directory expression
 
-            return (first locate expression)
+{- | Like `interpretWith`, but takes a pre-parsed expression instead of some
+     source code
+-}
+interpretExprWith
+    :: (MonadError InterpretError m, MonadIO m)
+    => [(Text, Type Location, Value)]
+    -- ^ @(name, type, value)@ for each custom binding
+    -> Maybe (Type Location)
+    -- ^ Optional expected type for the input
+    -> FilePath
+    -- ^ The base directory used when resolving filepath imports of relative paths
+    -> Syntax Location Import
+    -> m (Type Location, Value)
+interpretExprWith bindings maybeAnnotation directory expression = do
+    resolvedExpression <- for (annotate expression)
+        \(maybeAnnotation', input) -> case input of
+            Import.File file -> interpretWith bindings maybeAnnotation' (Path path)
+              where
+                path = FilePath.normalise (directory </> file)
+
+            Import.URI uri -> do
+                eitherResult <- liftIO do
+                    -- CUSTOMIZE ME
+                    -- If you want support for other URI types you probably want
+                    -- to replace the default resolver with your own one.
+                    tryAny (Import.resolverToCallback Resolver.defaultResolver uri)
+
+                importExpression <- case eitherResult of
+                    Left e -> throwError (ImportError uri (displayException e))
+                    Right result -> return result
+
+                let relocate location = location
+                        { name = Text.unpack (URI.render uri)
+                        }
 
-    let resolve (maybeAnnotation', file) =
-            interpretWith bindings maybeAnnotation' (Path path)
-          where
-            path = case input of
-                Path parent -> FilePath.takeDirectory parent </> file
-                Code _      -> file
+                let relocatedExpression = first relocate importExpression
 
-    resolvedExpression <- traverse resolve (annotate expression)
+                interpretExprWith bindings maybeAnnotation' directory relocatedExpression
```

Comment:
There's a subtle issue here, which is that if the user provides a `file://` URI then the `directory` won't be updated, so transitive imports will not be anchored correctly

I think the easiest way to fix this would be for the `Resolver` type to change to something like:

```haskell
URI -> IO (Maybe (Syntax Location Import, Maybe FilePath))
```

… so that it can optionally update the base directory.

Reply:
Done.